File: modules/exec-bridge/android/src/main/java/dev/chelly/execbridge/ExecBridgeModule.kt
package dev.chelly.execbridge

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.io.File

class ExecBridgeModule : Module() {
    override fun definition() = ModuleDefinition {
        Name("ExecBridge")

        AsyncFunction("execCommand") { command: String, cwd: String, timeoutMs: Int ->
            val linkerPath = findLinker()
            val bashPath = findBash()
            val ldLibPath = findLdLibPath()
            val filesDir = appContext.reactContext?.filesDir
                ?: throw RuntimeException("React context filesDir unavailable")
            val homePath = File(filesDir, "home").absolutePath
            File(homePath).mkdirs()
            val workDir = if (cwd.isNotEmpty()) cwd else "$homePath/chelly/workspace"
            File(workDir).mkdirs()

            val result = ChellyJNI.execSubprocess(
                linkerPath, bashPath, ldLibPath, homePath, workDir, command, timeoutMs
            )
            mapOf(
                "exitCode" to (result[0].toIntOrNull() ?: -1),
                "stdout" to result[1],
                "stderr" to result[2]
            )
        }
    }

    private fun findLinker(): String {
        val candidates = listOf(
            "/system/bin/linker64",
            "/apex/com.android.runtime/bin/linker64",
            "/system/bin/linker"
        )
        return candidates.firstOrNull { File(it).exists() }
            ?: throw RuntimeException("linker64 not found")
    }

    private fun findBash(): String {
        val appLib = appContext.reactContext?.applicationInfo?.nativeLibraryDir ?: ""
        val candidates = listOf("$appLib/libbash.so", "/system/bin/sh")
        return candidates.firstOrNull { File(it).exists() } ?: "/system/bin/sh"
    }

    private fun findLdLibPath(): String {
        return appContext.reactContext?.applicationInfo?.nativeLibraryDir ?: ""
    }
}

File: modules/exec-bridge/src/ExecBridgeModule.ts
import { requireNativeModule } from "expo-modules-core";
import { Platform } from "react-native";

export type ExecResult = {
  stdout: string;
  stderr: string;
  exitCode: number;
};

type ExecBridgeNative = {
  execCommand: (command: string, cwd: string, timeoutMs: number) => Promise<ExecResult>;
};

let cached: ExecBridgeNative | null = null;

function getNative(): ExecBridgeNative {
  if (cached) return cached;
  if (Platform.OS === "web") {
    cached = {
      async execCommand(): Promise<ExecResult> {
        return {
          stdout: "",
          stderr: "exec-bridge is not available on web",
          exitCode: -1,
        };
      },
    };
  } else {
    cached = requireNativeModule("ExecBridge") as ExecBridgeNative;
  }
  return cached;
}

export async function execCommand(
  command: string,
  cwd?: string,
  timeoutMs: number = 30000,
): Promise<ExecResult> {
  return getNative().execCommand(command, cwd ?? "", timeoutMs);
}
